### PR TITLE
feat(scaletest): annotate scaletest pod when scaletest is in progress

### PR DESCRIPTION
**Pull request by johnstcn**
Part of https://github.com/coder/coder/issues/7777

This PR modifies the scaletest-runner template to add a pod annotation to the scaletest runner pod. 

The annotation key is set to `com.coder.scaletest.phase` and the annotation value is one of `preparing`, `running`, or `completed`.

This will allow checking if a scaletest is in progress, and preventing any operations that would interrupt a running scaletest.

### CODE REVIEW

File: scaletest/templates/scaletest-runner/scripts/lib.sh [hunk 1]
```diff
@@ -247,37 +247,62 @@ set_appearance() {
 		"${CODER_URL}/api/v2/appearance"
 }
 
+namespace() {
+	local ns
+	ns=$(</var/run/secrets/kubernetes.io/serviceaccount/namespace)
+	echo "${ns}"
```
Comment (mafredri) [inline scaletest/templates/scaletest-runner/scripts/lib.sh:253]: A plain `cat` seems simpler here?

File: scaletest/templates/scaletest-runner/scripts/lib.sh [hunk 2]
```diff
@@ -247,37 +247,62 @@ set_appearance() {
...
+}
+coder_pods() {
+	local pods
+	pods=$(kubectl get pods \
+		--namespace "$(namespace)" \
+		--selector "app.kubernetes.io/name=coder,app.kubernetes.io/part-of=coder" \
+		--output jsonpath='{.items[*].metadata.name}')
+	echo "${pods}"
```
Comment (mafredri) [inline scaletest/templates/scaletest-runner/scripts/lib.sh:261]: Do we need the intermediary variable?

File: scaletest/templates/scaletest-runner/scripts/lib.sh [hunk 3]
```diff
@@ -247,37 +247,62 @@ set_appearance() {
...
-	log "Namespace from serviceaccount token is ${namespace}"
-	pod=$(kubectl get pods \
-		--namespace "${namespace}" \
-		--selector "app.kubernetes.io/name=coder,app.kubernetes.io/part-of=coder" \
-		--output jsonpath='{.items[0].metadata.name}')
+	log "Namespace from serviceaccount token is ${ns}"
+	pods=$(coder_pods)
+	if [[ -z ${pods} ]]; then
+		log "Could not find coder pods!"
+		exit 1
```
Comment (mafredri) [inline scaletest/templates/scaletest-runner/scripts/lib.sh:280]: We should only use `return` in all library functions, this allows the calling scripts to still handle errors. Using `exit` terminates the entire script immediately.

I see there's also a bug in `end_phase` where we use `exit` too.

File: scaletest/templates/scaletest-runner/scripts/lib.sh [hunk 4]
```diff
@@ -247,37 +247,62 @@ set_appearance() {
...
+	fi
+	pod=$(echo "${pods}" | cut -d ' ' -f 1)
```
Comment (mafredri) [inline scaletest/templates/scaletest-runner/scripts/lib.sh:282]: No need for echo here:
```suggestion
	pod=$(cut -d ' ' -f 1 <<<"${pods}")
```

File: scaletest/templates/scaletest-runner/scripts/lib.sh [hunk 5]
```diff
@@ -247,37 +247,62 @@ set_appearance() {
...
 		--container coder \
 		--retries 10 \
 		"${pod}:/opt/coder" "${SCALETEST_CODER_BINARY}"
 	maybedryrun "${DRY_RUN}" chmod +x "${SCALETEST_CODER_BINARY}"
 	log "Full Coder binary downloaded to ${SCALETEST_CODER_BINARY}"
 }
+
+# set_status_annotation annotates the currently running pod with the key
+# com.coder.scaletest.status. It will overwrite the previous status.
+set_status_annotation() {
```
Comment (mafredri) [inline scaletest/templates/scaletest-runner/scripts/lib.sh:301]: Maybe we can rename this to `set_pod_status[_annotation]`? Should we also call it automatically from `set_status` so that it's always up-to-date with scale test status? If you need lower case you can do `tr '[A-Z]' '[a-z]' <<<"$1"`.

```sh
❯ tr '[A-Z]' '[a-z]' <<<Preparing
preparing
```

Comment (johnstcn) [reply scaletest/templates/scaletest-runner/scripts/lib.sh:301]: That's a good idea!

File: scaletest/templates/scaletest-runner/scripts/lib.sh (hunk 2 above)
Comment (johnstcn) [reply scaletest/templates/scaletest-runner/scripts/lib.sh]: Probably not!

File: scaletest/templates/scaletest-runner/scripts/lib.sh (hunk 1 above)
Comment (johnstcn) [reply scaletest/templates/scaletest-runner/scripts/lib.sh]: 🐱

File: scaletest/templates/scaletest-runner/scripts/lib.sh [hunk 6]
```diff
@@ -40,7 +40,7 @@ show_json() {
 set_status() {
 	dry_run=
 	if [[ ${DRY_RUN} == 1 ]]; then
-		dry_run=" (dry-ryn)"
+		dry_run=" (dry-run)"
```
Comment (johnstcn) [inline scaletest/templates/scaletest-runner/scripts/lib.sh:43]: drive-by 😁

File: scaletest/templates/scaletest-runner/scripts/lib.sh [hunk 7]
```diff
@@ -271,18 +270,18 @@ fetch_coder_full() {
 	ns=$(namespace)
 	if [[ -z "${ns}" ]]; then
 		log "Could not determine namespace!"
-		exit 1
+		return
```
Comment (mafredri) [inline scaletest/templates/scaletest-runner/scripts/lib.sh:273]: We can still indicate error here by returning a non-zero value, otherwise the return code of `log` will be implicitly used (which is `0`).

```suggestion
		return 1
```

**Review by mafredri — APPROVED** [review]
